File: stream-chat-android-docs/src/main/java/io/getstream/chat/docs/java/Moderation.java
package io.getstream.chat.docs.java;

import java.util.Date;
import java.util.List;

import io.getstream.chat.android.client.ChatClient;
import io.getstream.chat.android.client.api.models.FilterObject;
import io.getstream.chat.android.client.api.models.QuerySort;
import io.getstream.chat.android.client.api.models.QueryUsersRequest;
import io.getstream.chat.android.client.channel.ChannelClient;
import io.getstream.chat.android.client.models.BannedUser;
import io.getstream.chat.android.client.models.BannedUsersSort;
import io.getstream.chat.android.client.models.Filters;
import io.getstream.chat.android.client.models.Flag;
import io.getstream.chat.android.client.models.Member;
import io.getstream.chat.android.client.models.Mute;
import io.getstream.chat.android.client.models.User;

import static java.util.Collections.emptyList;

public class Moderation {
    private ChatClient client;
    private ChannelClient channelClient;

    /**
     * @see <a href="hhttps://getstream.io/chat/docs/android/moderation/?language=java#flag">Flag</a>
     */
    class Flags {

        public void flagMessage() {
            client.flagMessage("message-id").enqueue(result -> {
                if (result.isSuccess()) {
                    // Message was flagged
                    Flag flag = result.data();
                } else {
                    // Handle result.error()
                }
            });
        }

        public void unflagMessage() {
            client.unflagMessage("message-id").enqueue(result -> {
                if (result.isSuccess()) {
                    // Message flag was removed
                    Flag flag = result.data();
                } else {
                    // Handle result.error()
                }
            });
        }

        public void flagUser() {
            client.flagUser("user-id").enqueue(result -> {
                if (result.isSuccess()) {
                    // User was flagged
                    Flag flag = result.data();
                } else {
                    // Handle result.error()
                }
            });
        }

        public void unflagUser() {
            client.unflagUser("user-id").enqueue(result -> {
                if (result.isSuccess()) {
                    // User flag was removed
                    Flag flag = result.data();
                } else {
                    // Handle result.error()
                }
            });
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/moderation/?language=java#mutes">Mutes</a>
     */
    class Mutes {

        public void muteUser() {
            client.muteUser("user-id").enqueue(result -> {
                if (result.isSuccess()) {
                    // User was muted
                    Mute mute = result.data();
                } else {
                    // Handle result.error()
                }
            });
        }

        public void unmuteUser() {
            client.unmuteUser("user-id").enqueue(result -> {
                if (result.isSuccess()) {
                    // User was unmuted
                } else {
                    // Handle result.error()
                }
            });
        }
    }


    /**
     * @see <a href="https://getstream.io/chat/docs/android/moderation/?language=java#ban">Bans</a>
     */
    class Bans {

        public void banUser() {
            // Ban user for 60 minutes from a channel
            channelClient.banUser("user-id", "Bad words", 60).enqueue(result -> {
                if (result.isSuccess()) {
                    // User was banned
                } else {
                    // Handle result.error()
                }
            });
        }

        public void unbanUser() {
            channelClient.unbanUser("user-id").enqueue(result -> {
                if (result.isSuccess()) {
                    // User was unbanned
                } else {
                    // Handle result.error()
                }
            });
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/moderation/?language=java#list-banned-users">List banned users</a>
     */
    class ListBannedUsers {

        public void queryBannedUsers() {
            // Retrieve the list of banned users
            FilterObject filter = Filters.eq("banned", true);
            QueryUsersRequest request = new QueryUsersRequest(filter, 0, 10);
            client.queryUsers(request).enqueue(result -> {
                if (result.isSuccess()) {
                    List<User> users = result.data();
                } else {
                    // Handle result.error()
                }
            });

            // Query for banned members from one channel
            FilterObject channelFilter = Filters.eq("channel_cid", "ChannelType:ChannelId");
            client.queryBannedUsers(channelFilter).enqueue(result -> {
                if (result.isSuccess()) {
                    List<BannedUser> bannedUsers = result.data();
                } else {
                    // Handle result.error()
                }
            });
        }

        public void queryBannedMembers() {
            FilterObject filter = Filters.eq("banned", true);
            QuerySort<Member> sort = new QuerySort<>();
            channelClient.queryMembers(0, 10, filter, sort, emptyList()).enqueue(result -> {
                if (result.isSuccess()) {
                    List<Member> members = result.data();
                } else {
                    // Handle result.error()
                }
            });
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/react/moderation/?language=java#query-bans-endpoint">Query bans endpoint</a>
     */
    class QueryBansEndpoint {

        public void queryBans() {
            // Get the bans for channel livestream:123 in descending order
            FilterObject filter = Filters.eq("channel_cid", "livestream:123");
            QuerySort<BannedUsersSort> sort = new QuerySort<BannedUsersSort>().desc("created_at");
            client.queryBannedUsers(filter, sort).enqueue(result -> {
                if (result.isSuccess()) {
                    List<BannedUser> bannedUsers = result.data();
                } else {
                    // Handle result.error()
                }
            });

            // Get the page of bans which where created before or equal date for the same channel
            client.queryBannedUsers(filter, sort, null, null, null, null, null, new Date()).enqueue(result -> {
                if (result.isSuccess()) {
                    List<BannedUser> bannedUsers = result.data();
                } else {
                    // Handle result.error()
                }
            });
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/moderation/?language=java#shadow-ban">Shadow ban</a>
     */
    class ShadowBan {

        public void shadowBanUser() {
            // Shadow ban user for 60 minutes from a channel
            channelClient.shadowBanUser("user-id", "Bad words", 60).enqueue(result -> {
                if (result.isSuccess()) {
                    // User was shadow banned
                } else {
                    // Handle result.error()
                }
            });
        }

        public void removeShadowBan() {
            channelClient.removeShadowBan("user-id").enqueue(result -> {
                if (result.isSuccess()) {
                    // Shadow ban was removed
                } else {
                    // Handle result.error()
                }
            });
        }
    }
}

File: stream-chat-android-docs/src/main/kotlin/io/getstream/chat/docs/kotlin/Moderation.kt
package io.getstream.chat.docs.kotlin

import io.getstream.chat.android.client.ChatClient
import io.getstream.chat.android.client.api.models.QuerySort
import io.getstream.chat.android.client.api.models.QueryUsersRequest
import io.getstream.chat.android.client.channel.ChannelClient
import io.getstream.chat.android.client.models.BannedUser
import io.getstream.chat.android.client.models.BannedUsersSort
import io.getstream.chat.android.client.models.Filters
import io.getstream.chat.android.client.models.Flag
import io.getstream.chat.android.client.models.Member
import io.getstream.chat.android.client.models.Mute
import io.getstream.chat.android.client.models.User
import java.util.Date

class Moderation(val client: ChatClient, val channelClient: ChannelClient) {

    /**
     * @see <a href="hhttps://getstream.io/chat/docs/android/moderation/?language=kotlin#flag">Flag</a>
     */
    inner class Flags {

        fun flagMessage() {
            client.flagMessage("message-id").enqueue { result ->
                if (result.isSuccess) {
                    // Message was flagged
                    val flag: Flag = result.data()
                } else {
                    // Handle result.error()
                }
            }
        }

        fun unflagMessage() {
            client.unflagMessage("message-id").enqueue { result ->
                if (result.isSuccess) {
                    // Message flag was removed
                    val flag: Flag = result.data()
                } else {
                    // Handle result.error()
                }
            }
        }

        fun flagUser() {
            client.flagUser("user-id").enqueue { result ->
                if (result.isSuccess) {
                    // User was flagged
                    val flag: Flag = result.data()
                } else {
                    // Handle result.error()
                }
            }
        }

        fun unflagUser() {
            client.unflagUser("user-id").enqueue { result ->
                if (result.isSuccess) {
                    // User flag was removed
                    val flag: Flag = result.data()
                } else {
                    // Handle result.error()
                }
            }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/moderation/?language=kotlin#mutes">Mutes</a>
     */
    inner class Mutes {

        fun muteUser() {
            client.muteUser("user-id").enqueue { result ->
                if (result.isSuccess) {
                    // User was muted
                    val mute: Mute = result.data()
                } else {
                    // Handle result.error()
                }
            }
        }

        fun muteUserWithTimeout() {
            // Mute user for 60 minutes
            client.muteUser("user-id", timeout = 60).enqueue { result ->
                if (result.isSuccess) {
                    // User was muted
                } else {
                    // Handle result.error()
                }
            }
        }

        fun unmuteUser() {
            client.unmuteUser("user-id").enqueue { result ->
                if (result.isSuccess) {
                    // User was unmuted
                } else {
                    // Handle result.error()
                }
            }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/moderation/?language=kotlin#ban">Bans</a>
     */
    inner class Bans {

        fun banUser() {
            // Ban user for 60 minutes from a channel
            channelClient.banUser(targetId = "user-id", reason = "Bad words", timeout = 60).enqueue { result ->
                if (result.isSuccess) {
                    // User was banned
                } else {
                    // Handle result.error()
                }
            }
        }

        fun unbanUser() {
            channelClient.unbanUser(targetId = "user-id").enqueue { result ->
                if (result.isSuccess) {
                    // User was unbanned
                } else {
                    // Handle result.error()
                }
            }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/moderation/?language=kotlin#list-banned-users">List banned users</a>
     */
    inner class ListBannedUsers {

        fun queryBannedUsers() {
            // Retrieve the list of banned users
            client.queryUsers(
                QueryUsersRequest(
                    filter = Filters.eq("banned", true),
                    offset = 0,
                    limit = 10,
                )
            ).enqueue { result ->
                if (result.isSuccess) {
                    val users: List<User> = result.data()
                } else {
                    // Handle result.error()
                }
            }

            // Query for banned members from one channel
            client.queryBannedUsers(filter = Filters.eq("channel_cid", "ChannelType:ChannelId")).enqueue { result ->
                if (result.isSuccess) {
                    val bannedUsers: List<BannedUser> = result.data()
                } else {
                    // Handle result.error()
                }
            }
        }

        fun queryBannedMembers() {
            channelClient.queryMembers(
                offset = 0,
                limit = 10,
                filter = Filters.eq("banned", true),
                sort = QuerySort(),
                members = emptyList()
            ).enqueue { result ->
                if (result.isSuccess) {
                    val members: List<Member> = result.data()
                } else {
                    // Handle result.error()
                }
            }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/react/moderation/?language=kotlin#query-bans-endpoint">Query bans endpoint</a>
     */
    inner class QueryBansEndpoint {

        fun queryBans() {
            // Get the bans for channel livestream:123 in descending order
            client.queryBannedUsers(
                filter = Filters.eq("channel_cid", "livestream:123"),
                sort = QuerySort.desc(BannedUsersSort::createdAt),
            ).enqueue { result ->
                if (result.isSuccess) {
                    val bannedUsers: List<BannedUser> = result.data()
                } else {
                    // Handle result.error()
                }
            }

            // Get the page of bans which where created before or equal date for the same channel
            client.queryBannedUsers(
                filter = Filters.eq("channel_cid", "livestream:123"),
                sort = QuerySort.desc(BannedUsersSort::createdAt),
                createdAtBeforeOrEqual = Date(),
            ).enqueue { result ->
                if (result.isSuccess) {
                    val bannedUsers: List<BannedUser> = result.data()
                } else {
                    // Handle result.error()
                }
            }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/moderation/?language=kotlin#shadow-ban">Shadow ban</a>
     */
    inner class ShadowBan {

        fun shadowBanUser() {
            // Shadow ban user for 60 minutes from a channel
            channelClient.shadowBanUser(targetId = "user-id", reason = "Bad words", timeout = 60).enqueue { result ->
                if (result.isSuccess) {
                    // User was shadow banned
                } else {
                    // Handle result.error()
                }
            }
        }

        fun removeShadowBan() {
            channelClient.removeShadowBan("user-id").enqueue { result ->
                if (result.isSuccess) {
                    // Shadow ban was removed
                } else {
                    // Handle result.error()
                }
            }
        }
    }
}
